spec.bs

```diff
@@ -461,6 +473,146 @@ An <dfn>internal observer</dfn> is a [=struct=] with the following [=struct/item
+
+       Note: This prevents primitive types from being coerced into iterables (e.g., String).
+
+       Issue: See if this is even the behavior we want. See <a
+       href=https://github.com/WICG/observable/issues/125>WICG/observable#125</a>
+
+    1. <i id=from-observable-conversion><b>From Observable</b></i>: If |value|'s [=specific type=]
+       is an {{Observable}}, then return |value|.
+
+    1. <i id=from-async-iterable-conversion><b>From async iterator</b></i>: Let
+       |asyncIteratorMethodRecord| be [=?=] [$GetMethod$](|value|, {{%Symbol.asyncIterator%}}).
+
+       Note: We have to use [$GetMethod$] directly instead of [$GetIterator$], because
+       [$GetIterator$] makes it impossible to distinguish between (a) the case where |value| simply
+       doesn't implement any iterable protocol, and (b) the case where |value| implements one of the
+       iterator protocols, and its method getter throws an exception.
+
+       Note: This step is web-observable, and re-throws any errors that the {{%Symbol.iterator%}}
+       method *getter* might have thrown.
+
+    1. If |asyncIteratorMethodRecord|'s \[[Value]] is undefined or null, then jump to the step
+       labeled <a href=#from-iterable-conversion>From iterable</a>.
+
+    1. If [$IsCallable$](|asyncIteratorMethodRecord|'s \[[Value]]) is false, then
+       [=exception/throw=] a {{TypeError}}.
+
+       Otherwise, let |nextAlgorithm| be the following steps, given |subscriber| and |iterator|:
+
+         1. If |iterator|'s \[[Done]] is true, then: run |subscriber|'s {{Subscriber/complete()}}
+            method and abort these steps.
+
+         1. Let |nextRecord| be [$IteratorStepValue$](|iterator|).
+
+         1. Let |nextPromise| be undefined.
+
+         1.  If |nextRecord| is a [=throw completion=], then set |nextPromise| to [=a promise
+             rejected with=] |nextRecord|'s \[[Value]].
+
+         1. Otherwise, set |nextPromise| to |nextRecord|'s \[[Value]].
+
+         1. [=React=] to |nextPromise|:
+
+            1. If |nextPromise| was fulfilled with value |v|, then:
+
+               1. Run |subscriber|'s {{Subscriber/next()}} method, given |v|.
+
+               1. Run |nextAlgorithm|, given |subscriber| and |iterator|.
+
+            1. If |nextPromise| was rejected with reason |r|, then run |subscriber|'s
+               {{Subscriber/error()}} method, given |r|.
+
+    1. Return a [=new=] {{Observable}} whose [=Observable/subscribe callback=] is an algorithm that
+       takes a {{Subscriber}} |subscriber| and does the following:
+
+       1. Let |iteratorRecord| be [$GetIteratorFromMethod$](|value|, {{%Symbol.asyncIterator%}}).
+
+       1. If |iteratorRecord| is a [=throw completion=], then [=queue a microtask=] to run
+          |subscriber|'s {{Subscriber/error()}} method, given |iteratorRecord|'s \[[Value]].
+
+       1. Otherwise, [=queue a microtask=] to run |nextAlgorithm| given |subscriber| and
```

Comment:
I don't think we want to queue this microtask here. I understood the original intent (so that in the event the iterator is `[[Done]]`, we run `complete()` asynchronously), but from looking at this more the IteratorRecord that `GetIterator()` always returns has `[[Done]]: false`, so I think we don't need to do this up-front checking at all. Furthermore, queuing this microtask has web-observable consequences.

With it, the invocation of `next()` is behind at least two microtasks. First, is the one we queue here, and second is the one is the one we get by "reacting" to the promise returned from the iterator's `[[NextMethod]]`. I think this is web-observable, because if the retrieval and/or invocation of `%Symbol.asyncIterator%` itself queues a microtask, this microtask would interleave the two aforementioned ones, which seems weird.

To simplify the structure of this, I think we should move the done check and `complete()` call after the `next()` method invocation (behind the reaction microtask, but still synchronous with respect to `subscriber->next()`). That way it's still queued behind a microtask, but it's in the right place.

Reply:
Done in https://github.com/WICG/observable/pull/160/commits/4a34553cc13fe1e7b06c33db5b02e8051dca71bf. Most of the above comment was right I think, but we don't want to move the done check to after we invoke `subscriber->next()`. I've restructured things a bit and added a note about microtask timing.